### DISCUSSION TITLE
How to refuse overwrite on INSERT for created_at `NOW()`

### DISCUSSION THREAD

**Discussion by bennik88** [supabase/supabase#9053] [Questions]
## Problem
I'm struggling to get a restriction that disallows overwriting the fields `creaded_at`, `id` and `owner_id` on INSERT. I got something to work when it comes to UPDATE with a function and triggers by checking the `OLD` and `NEW` values. But this doesn't seem to work on INSERT.

## What I've tried
 It seems that `NOW()` isn't populated before the TRIGGER runs. This is weird as the user can overwrite `NOW()` but in TRIGGER, BEFORE INSERT it seems to be not yet available. 
```sql
CREATE OR REPLACE FUNCTION validate_insert()
RETURNS TRIGGER AS
$$
BEGIN
  -- always reset id created_at and owner_id to the value already stored
  NEW.id := OLD.id;
  NEW.created_at := OLD.created_at;
  NEW.owner_id := OLD.owner_id;
  RETURN NEW;
END;
$$
LANGUAGE plpgsql;
```

I've also tried it using this, which also doesn't work. The user can still overwrite the three fields:
```sql
BEGIN
  IF NEW.id <> NULL
  OR NEW.created_at <> NULL
  OR NEW.last_updated_at <> NULL
   THEN
    RAISE EXCEPTION 'Cannot set id, created_at or last_updated_at';
    END IF;
    RETURN NEW;
END;
```

I'm also open for any other solutions to approach this. As I will need this for every table. Anything I can easily adapt to new tables is welcome.

**Comment by GaryAustin1** [comment]
You can just use new.created_at = now() with return new.

You can also handle both insert and update in the same function by using TG_OP to do different things.  https://www.postgresql.org/docs/current/plpgsql-trigger.html

**Reply by bennik88** [reply]
1. Using `NEW.created_at = now()` makes the Dashboard setting for the column Default Value `now()` obsolete right? 
I therefore could just set Not Nullable in the Supabase UI as the `now()` value would never be used. 

In the PostgreSQL docs I found that they check in trigger functions weather a value is given or not 
```sql
-- Check that empname and salary are given
        IF NEW.empname IS NULL THEN
            RAISE EXCEPTION 'empname cannot be null';
        END IF;
```
2. Is this considered a best practice when working with Postgres/ Supabase even when table columns are created with `NOT NULL?

The order of operations would be interesting to know. I'm very confused by the `NEW.value` not being overwritten by the `OLD.value` on INSERT. Till now my INSERT and UPDATE triggers have had the activation `BEFORE`. 
Correct me if I'm wrong but my conclusion would be the below. 
3. Is that right?

**INSERT**
1. The `INSERT` event calls the default value `now()` (and it seems to not go into the `OLD` var)
2. Then the trigger could overwrite the data if given. (OLD is not accessible, I don't know why. I expected now() to be in the OLD var)
3. The value given (by the user who made the request) with the `INSERT` statement would overwrite all other vars from before. 
4. Everything is saved to the table.

**UPDATE**
1. The `UPDATE` event gets the OLD row values.
2. The TRIGGER BEFORE runs and updates the NEW values with the OLD values.
3. Everything is saved to the table.

4. Would I, therefore, have to use an `AFTER` trigger?

Also, thank you for the hint with `NEW.created_at = now()`.
5. What should I do about the id? Should I generate the id in the trigger function then like `uuid_generate_v4()` or should I go for the below?
```sql
IF NEW.id IS NOT NULL THEN
            RAISE EXCEPTION 'id cannot be set';
```

Here is btw where I'm at thanks to your hint to the `TG_OP` var. Sorry for all the questions I'm giving my best to understand how to work with PostgreSQL to make Supabase work.

6. Do I have to use a `BEFORE` trigger or should I use the `AFTER` trigger to really make sure the latest overwrites are made by the trigger?
```sql
CREATE OR REPLACE FUNCTION process_contacts_audit() RETURN TRIGGER AS $$
  BEGIN
    IF (TG_OP = 'INSERT') THEN
      -- Code for INSERT
      NEW.id = uuid_generate_v4();
      NEW.created_at = now();
      NEW.last_updated_at = now();
      RETURN NEW;
    ELSIF (TG_OP = 'UPDATE') THEN
      -- Code for UPDATE
      NEW.id = OLD.id;
      NEW.created_at = OLD.created_at;
      NEW.owner_id = OLD.owner_id;
      NEW.last_updated_at = now();
      RETURN NEW;
    RETURN NULL; -- result is ignored since this is an AFTER trigger
  END;
$$ LANGUAGE plpgsql;

CREATE TRIGGER contacts_audit
AFTER INSERT OR UPDATE ON contacts
  FOR EACH ROW EXECUTE FUNCTION process_contacts_audit();
```

**Reply by bennik88** [reply]
@GaryAustin1 Any idea?
